### ISSUE TITLE
Questions about the criteria for choosing the final A_distance

### ISSUE THREAD

**Issue by k0ngyiji** [jindongwang/transferlearning#199]
Hi, 
I come into trouble when I compute the A-distance for the representation learned by my model. 
I find that the final accuracy has no certain relationship with A-distance, e.g. higher accuracy corresponding to lower A-distance. In  some of my experiments, some tasks with higher accuracy get worse A-distance. 
Besides, I reproduced the DSAN and compute the final A-distance on task A-W every epoch, here is my result:
![image](https://user-images.githubusercontent.com/68325219/96684694-95006380-13ae-11eb-9e66-0098e820dad3.png)
which has different A-distance with the original paper:
![image](https://user-images.githubusercontent.com/68325219/96684564-5d91b700-13ae-11eb-847b-b69e51545008.png)
So I think there is something wrong with the way I compute the A-distance. I use the code in https://github.com/jindongwang/transferlearning/blob/4a00b774dd6d86e19552d5964c68ee005514c763/code/distance/proxy_a_distance.py. 
Best wishes!!!!!!!!

**Comment by jindongwang** [comment]
A-distance is not the ideal distance for computing domain distances. So, its results may vary according to your algorithms. May I ask which algorithm you use?

**Comment by k0ngyiji** [comment]
Thank you for your response !!
Inspired by DSAN, I come up with a new idea and try to reproduce the results in DSAN.  If the AL_distance is Interpretable, I will cite it in my paper. 
However, it seems that no paper gives some details on how to choose the final A_distance, so it also confuses me on how to choose the proper AL_distance.
Can you give me some suggestions? Here is my code for computing A_distance and AL_distance:
def proxy_distance(model,p_c):
    import numpy as np
    from sklearn import svm
    from numpy import linalg as LA
    def proxy_a_distance(source_X, target_X, verbose=False):


        """
        Compute the Proxy-A-Distance of a source/target representation
        """
        source_X = source_X.detach().cpu().numpy()
        target_X = target_X.detach().cpu().numpy()
        nb_source = np.shape(source_X)[0]
        nb_target = np.shape(target_X)[0]


        if verbose:
            print('PAD on', (nb_source, nb_target), 'examples')


        C_list = np.logspace(-5, 4, 10)


        half_source, half_target = int(nb_source/2), int(nb_target/2)
        train_X = np.vstack((source_X[0:half_source, :], target_X[0:half_target, :]))
        train_Y = np.hstack((np.zeros(half_source, dtype=int), np.ones(half_target, dtype=int)))


        test_X = np.vstack((source_X[half_source:, :], target_X[half_target:, :]))
        test_Y = np.hstack((np.zeros(nb_source - half_source, dtype=int), np.ones(nb_target - half_target, dtype=int)))


        best_risk = 1.0
        for C in C_list:
            clf = svm.SVC(C=C, kernel="rbf" verbose=False)
            clf.fit(train_X, train_Y)


            train_risk = np.mean(clf.predict(train_X) != train_Y)
            test_risk = np.mean(clf.predict(test_X) != test_Y)


            if verbose:
                print('[ PAD C = %f ] train risk: %f  test risk: %f' % (C, train_risk, test_risk))


            if test_risk > .5:
                test_risk = 1. - test_risk


            best_risk = min(best_risk, test_risk)
        return 2 * (1. - 2 * best_risk)


    ######pseudo-code used to extract features from deep model
    source_feature = model.forward(source_dataloader)
    target_feature = model.forward(target_dataloader)
    A_distance = proxy_a_distance(source_feature,target_feature)
    Al_distance = 0
    for i in torch.unique(ys):
        Al_distance += p_c[i.item()]*proxy_a_distance(source_feature[ys==i.item(),:],target_feature[yt==i.item(),:])
    return A_distance,Al_distance
| |
kongyiji_is_coding
|
|
kongyiji_is_coding@163.com
|
签名由网易邮箱大师定制
On 10/24/2020 12:37，Jindong Wang<notifications@github.com> wrote：

A-distance is not the ideal distance for computing domain distances. So, its results may vary according to your algorithms. May I ask which algorithm you use?

**Comment by jindongwang** [comment]
@k0ngyiji Gong through your code and can't find anything wrong. I think that maybe you can increase your train_X and train_Y data size to see if it is right. And you can also try SVM with linear kernel. Free to discuss further changes.

**Comment by k0ngyiji** [comment]
I have tried every kernel before. Considering about the cosine similarity in my method, I define a cosine similarity kernel, which actually improves the A-distance greatly. This finding proves that the kernel function has impact on the A-distance. Here is my code:

from numpy import linalg as LA
def my_kernel(X, Y):
    norm = LA.norm(X) * LA.norm(Y)
    return np.dot(X, Y.T)/norm
clf = svm.SVC(C=C, **kernel=my_kernel,** verbose=False)

In our paper, should I use the same kernel function to compute the A-distance for different baselines reprouced by us, or directly use the A-distance displayed in their original paper? The former may led to the great difference between our reproduced results and the results in orginal paper.

**Comment by jindongwang** [comment]
It is OK to use any form of computation as long as it supports your analysis and prove that your algorithm is better.

**Comment by k0ngyiji** [comment]
I get it, and this actually confuses me for a long time. This really helps me a lot.
Thank you for your patience !!!

**Comment by khan-yin** [comment]
> I get it, and this actually confuses me for a long time. This really helps me a lot. Thank you for your patience !!!
@jindongwang @h0ngxuanli 

Hi! I got the same problem with you! I want to ask if I need to keep the same number of train and test dataset, I used all training domains data and only 1 test domain, it facing a large class balance problem. 
**I think it also affect its performance should I made training data the same size with test?** 
**or if a MLP classifer would be better, cause the SVM fitting is too long and can not handle different hidden_size features from different model**
